input-profile.ts

```diff
@@ -0,0 +1,322 @@
+    }
+
+    setComponentsActive(active: boolean) {
+        const comps = this.defaultControllerComponents;
+        if (!comps) return;
+        for (let i = 0; i < comps.length; ++i) {
+            comps[i].active = active;
+        }
+    }
+    // Save data to cache
+    saveToCache(key, data) {
+        try {
+            // Convert data to JSON string before storing
+            const jsonString = JSON.stringify(data);
+
+            // Save to local storage
+            localStorage.setItem(key, jsonString);
+        } catch (error) {
+            console.error('Error saving to cache:', error);
+        }
+    }
+
+    // Retrieve data from cache
+    retrieveFromCache(key) {
+        try {
+            // Retrieve data from local storage
+            const jsonString = localStorage.getItem(key);
+
+            // Parse JSON string to get the original data
+            const data = JSON.parse(jsonString);
+
+            return data;
+        } catch (error) {
+            console.error('Error retrieving from cache:', error);
+            return null;
+        }
+    }
+
+    onInputSourcesChange(event: any) {
+        if (this.modelLoaded && !this.mapToDefaultController) {
+            this.setComponentsActive(false);
+        }
+
+        event.added.forEach((xrInputSource: any) => {
+            if (xrInputSource.hand != null) return;
+            if (this.handedness != xrInputSource.handedness) return;
+            var profile = this.path + xrInputSource.profiles[0];
+            if (this.customProfileFolder !== '') profile = this.customProfileFolder;
+            this.url = profile + '/profile.json';
+            this.ProfileJSON = this.retrieveFromCache(this.url);
+            if (this.ProfileJSON != null) {
+                console.log('Loaded Profile From Cache ');
+                this.loadAndMapGamepad(profile, xrInputSource);
+            } else {
+                fetch(this.url)
+                    .then((res) => res.json())
+                    .then((out) => {
+                        this.ProfileJSON = out;
+                        console.log('Profile downloaded and loaded from ' + this.url);
+                        this.saveToCache(this.url, out);
+                        this.loadAndMapGamepad(profile, xrInputSource);
+                    })
+                    .catch((e) => {
+                        console.error('failed to load profile from' + this.url);
+                        console.error(e);
+                    });
+            }
+        });
+    }
+
+    isModelLoaded() {
+        return this.controllerModel !== null;
+    }
+
+    loadAndMapGamepad(profile: string, xrInputSource: any) {
+        this.gamepad = xrInputSource.gamepad;
+
+        const assetPath = profile + '/' + this.handedness + '.glb';
+        if (this.modelLoaded) return;
+
+        if (!this.mapToDefaultController) {
+            /** load 3d model in the runtime with profile url */
+            this.engine.scene
```

Comment:
This entire code can be summed up like that:

```ts
async loadAndMapGamepad(profile: string, xrInputSource: any) {
    this.gamepad = xrInputSource.gamepad;

    const assetPath = profile + '/' + this.handedness + '.glb';

    this.controllerModel = this.defaultController;
    if (!this.mapToDefaultController) {
        /** load 3d model in the runtime with profile url */
        try {
            this.controllerModel = await this.engine.scene.append(assetPath);
        } catch (e) {
            console.error('failed to load 3d model');
            console.error(e);
            this.setComponentsActive(true);
            console.log(
                'Couldnot load i-p, continuing with ' +
                    this.handedness +
                    ' default controller'
            );
        }
        this.setComponentsActive(false);
        console.log('Disabling ' + this.handedness + ' default Controller');
        this.controllerModel.parent = this.object;
        this.controllerModel.setTranslationLocal([0, 0, 0]);
        console.log(this.handedness + 'controller model loaded to the scene');
    } else {
        console.log('mapping i-p to ' + this.handedness + ' default controllers');
    }
    this.getGamepadObjectsFromProfile(this.ProfileJSON, this.controllerModel);
    this.setHandTrackingControllers(this.controllerModel);
    this.update = () => this.mapGamepadInput();
}
```

Much easier to read, and doesn't duplicate the object / hand setup

Reply:
any reason to put 
```js
this.controllerModel.parent = this.object;
this.controllerModel.setTranslationLocal([0, 0, 0]);
```
outside the try{} ?

Reply:
made the changes :thumbsup: